File: template/1a.py

#!/usr/bin/env python3

''' Performs Gibbs sampling on the specified set of sequences to identify motifs

Arguments:
    -f: set of sequences to identify a motif in
    -k: the length of the desired motif to find
Outputs:
    - a list of start positions for each sequence, along with the corresponding
      motif identified
    - the consensus motif (based on majority base at each position)
    - the position probability matrix

Example Usage:
    python3 1a.py -f motif1.fasta -k 10
'''

import argparse
from copy import deepcopy
import random
import numpy as np
from math import exp

np.random.seed(2950)

def sumLogProbs(a, b):
    if a > b:
        return a + np.log1p(exp(b-a))
    else:
        return b + np.log1p(exp(a-b))


''' Reads in the sequences from the motif files.

Arguments:
    filename: which filename to read sequences from
Returns:
    output: list of sequences in motif file
'''


def read_fasta(filename):
    with open(filename, "r") as f:
        output = []
        s = ""
        for l in f.readlines():
            if l.strip()[0] == ">":
                # skip the line that begins with ">"
                if s == "":
                    continue
                output.append(s)
                s = ""
            # keep appending to the current sequence when the line doesn't begin
            # with ">"
            else:
                s += l.strip()
        output.append(s)
        return output


''' Returns the majority vote (consensus) motif sequence.

Arguments:
    motif_sequences: list of all motif instances (list of strings)
Returns:
    majority sequence: consensus motif sequence (string)
'''


def majority(motif_sequences):
    base_ordering = {c: i for i, c in enumerate("ACGT")}
    freqs = np.zeros((len(motif_sequences[0]), 4))
    for m_s in motif_sequences:
        for i, b in enumerate(m_s):
            freqs[i, base_ordering[b]] += 1
    return ''.join(['ACGT'[np.argmax(row)] for row in freqs])


ACGT = {'A': 0, 'C': 1, 'G': 2, 'T': 3}

'''Adds the characters in a motif to an overall distribution array.    
Arguments:
    pi: the distribution array to increment
    sequence: the string containing the motif
    start: the start point
    k: the motif length
    add_subtr: whether to add or subtract 1 from pi
Returns:
    none
'''


def add_motif(pi, sequence, start, k, add_subtr):
    for i in range(0, k):
        current_char = sequence[start + i]
        pi[i][ACGT[current_char]] += add_subtr


'''Calculate the probability of a motif given a model.
Arguments: 
    model: the motif probability matrix
    str: the string containing the motif
    start: the start point of the motif
    k: the length of the motif
Returns:
    pr: the probability of the motif given the model
'''


def motif_prob(model, str, start, k, denom):
    pr = 0
    # probability of a motif = product of probability of each character
    for offset in range(0, k):
        current_char = str[start + offset]
        pr += np.log(model[offset][ACGT[current_char]])
    pr -= (k * denom)
    return pr


'''Calculate the likelihood of a set of motifs given their model.
Arguments: 
    model: the motif probability matrix
    starts: the start points of the motifs
    k: the length of the motif
    sequences: the list of sequences for the motifs
Returns:
'''


def likelihood(model, starts, k, sequences, denom):
    pr = 0
    # no need for denom if it's going to be constant between comparisons anyways
    for i in range(0, len(sequences)):
        pr += motif_prob(model, sequences[i],
                         starts[i], k, denom)
    return pr


''' Performs Gibbs sampling to estimate the start positions of the motif in
    each sequence and the motif model.

Arguments:
    sequences: list of sequences to find motifs in
    k: length of motif
    epsilon: pseudocounts for each base
Returns:
    starts: list of start positions for each sequence 
    pi: corresponding motif model (incorporating pseudocounts).
'''


def gibbs_sampling(sequences, k, epsilon):
    # initialize random motif starts (1 for each sequence, uniform from 0 to seqlen - k)
    num_seqs = len(sequences)
    small_denom = np.log(num_seqs - 1 + 4 * epsilon)
    big_denom = np.log(num_seqs + 4 * epsilon)
    starts = np.zeros(num_seqs, dtype=int)
    # for i in range(0, len(starts)):
    #     rand = np.random.randint(0, len(sequences[i]) - k + 1)
    #     starts[i] = rand
    # initialize ACGT distributions for each position in the motif
    pi = np.full((k, 4), epsilon)

    for j in range(0, num_seqs):
        add_motif(pi, sequences[j], starts[j], k, 1)

    best_starts = np.copy(starts)
    best_model = np.copy(pi)
    best_likelihood = likelihood(
        best_model, best_starts, k, sequences, big_denom)
    time_since_change = 0

    motif_distr = []
    for i in range(0, num_seqs):
        motif_distr.append(np.zeros(len(sequences[i]) - k + 1))

    # while condition holds:
    # for each motif n, update based on relative frequency in other sequences:
    # subtract current motif characters from distribution add_motif(n, -1), draw new start from remainder
    # update starts with new start:
    # add new motifs back to distr (add_motif(n, 1))
    first = True
    while(time_since_change < 100):
        for i in range(0, num_seqs):
            curr_str = sequences[i]
            add_motif(pi, sequences[i], starts[i], k, -1)
            if(first):
                print(pi)
            current_distr = motif_distr[i]
            # draw new start from remainder
            for curr_start in range(0, len(current_distr)):
                current_distr[curr_start] = motif_prob(
                    pi, curr_str, curr_start, k, small_denom)
            if(first):
                print(current_distr)
            # normalize before drawing: x - logsum(x1, x2 ... xn)
            logsum = current_distr[0]
            for curr_start in range(1, len(current_distr)):
                logsum = sumLogProbs(logsum, current_distr[curr_start])
            for j in range(0, len(current_distr)):
                current_distr[j] = exp(current_distr[j] - logsum)
            if(first):
                print(current_distr)
                first = False
            new_start = np.random.choice(len(current_distr), p=current_distr)
            starts[i] = new_start
            add_motif(pi, sequences[i], starts[i], k, 1)
        new_likelihood = likelihood(
            pi, starts, k, sequences, big_denom)
        if (new_likelihood > best_likelihood):
            best_likelihood = new_likelihood
            best_starts = np.copy(starts)
            best_model = deepcopy(pi)
            time_since_change = 0
        time_since_change += 1
    return best_starts, best_model / (num_seqs + 4 * epsilon)


def main():
    parser = argparse.ArgumentParser(
        description='Estimate the start positions and motif model via Gibbs sampling.')
    parser.add_argument('-f', action="store", dest="f",
                        type=str, default='sec9.fasta')
    parser.add_argument('-k', action="store", dest="k", type=int, default=4)
    parser.add_argument('-epsilon', action="store",
                        dest="epsilon", type=float, default=0.5)
    args = parser.parse_args()
    sequences = read_fasta(args.f)
    k = args.k
    epsilon = args.epsilon

    starts, pi = gibbs_sampling(sequences, k, epsilon)
    motif_sequences = [s[starts[i]:starts[i]+k]
                       for i, s in enumerate(sequences)]
    print('\n'.join([("Sequence %d: " % i) +
          m for i, m in enumerate(motif_sequences)]))
    print("Consensus motif: %s" % majority(motif_sequences))
    print(pi)


if __name__ == '__main__':
    main()

File: template/1a.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define MAX_SEQ_LEN 2000
#define MAX_SEQS 100
#define ACGT_CHARS "ACGT"
const int base_ordering[20] = {0,  -1, 1,  -1, -1, -1, 2,  -1, -1, -1,
                               -1, -1, -1, -1, -1, -1, -1, -1, -1, 3};

double sumLogProbs(double a, double b) {
    if (a > b) {
        return a + log1p(exp(b - a));
    } else {
        return b + log1p(exp(a - b));
    }
}

// Reads sequences from a FASTA file
int read_fasta(const char *filename, char sequences[MAX_SEQS][MAX_SEQ_LEN]) {
    FILE *file = fopen(filename, "r");
    if (!file) {
        perror("Error opening file");
        fprintf(stderr, "Failed to open: %s\n", filename);
        exit(EXIT_FAILURE);
    }
    char line[MAX_SEQ_LEN];
    int seq_count = 0;
    int index = 0;

    while (fgets(line, sizeof(line), file)) {
        if (line[0] == '>') {
            if (index > 0) {
                sequences[seq_count][index] = '\0';
                seq_count++;
                index = 0;
            }
        } else {
            char *pos = strchr(line, '\n');
            if (pos) *pos = '\0';  // Remove newline
            strcpy(&sequences[seq_count][index], line);
            index += strlen(line);
        }
    }
    sequences[seq_count][index] = '\0';
    fclose(file);
    return seq_count + 1;
}

// Computes the consensus motif from a set of motifs
void majority(char motif_sequences[MAX_SEQS][MAX_SEQ_LEN], int num_seqs,
              int motif_len, char *consensus) {
    int freqs[MAX_SEQ_LEN][4] = {0};
    int base_ordering[256] = {0};
    for (int i = 0; i < 4; i++) {
        base_ordering[(int)ACGT_CHARS[i]] = i;
    }

    for (int i = 0; i < num_seqs; i++) {
        for (int j = 0; j < motif_len; j++) {
            char base = motif_sequences[i][j];
            freqs[j][base_ordering[(int)base]]++;
        }
    }

    for (int i = 0; i < motif_len; i++) {
        int max_idx = 0;
        for (int j = 1; j < 4; j++) {
            if (freqs[i][j] > freqs[i][max_idx]) {
                max_idx = j;
            }
        }
        consensus[i] = ACGT_CHARS[max_idx];
    }
    consensus[motif_len] = '\0';
}

// Adds or subtracts a motif from the probability distribution
void add_motif(double pi[MAX_SEQ_LEN][4], const char *sequence, int start,
               int k, int add_subtr) {
    for (int i = 0; i < k; i++) {
        char base = sequence[start + i];
        pi[i][base_ordering[(int)base - 65]] += add_subtr;
    }
}

// Computes the probability of a motif
double motif_prob(double model[MAX_SEQ_LEN][4], const char *sequence, int start,
                  int k, double denom) {
    double pr = 0.0;
    for (int offset = 0; offset < k; offset++) {
        char base = sequence[start + offset];
        pr += log(model[offset][base_ordering[(int)base - 65]]);
    }
    pr -= (k * denom);
    return pr;
}

// Calculates the likelihood of a set of motifs given their model
double likelihood(double model[MAX_SEQ_LEN][4], int starts[MAX_SEQS],
                  int num_seqs, int k, char sequences[MAX_SEQS][MAX_SEQ_LEN],
                  double denom) {
    double pr = 0.0;

    // Sum the motif probabilities for each sequence
    for (int i = 0; i < num_seqs; i++) {
        pr += motif_prob(model, sequences[i], starts[i], k, denom);
    }

    return pr;
}

void print_pi(double pi[MAX_SEQ_LEN][4], int elements) {
    printf("-------\n");
    for (int i = 0; i < elements; i++) {
        for (int j = 0; j < 4; j++) {
            printf("%f ", pi[i][j]);
        }
        printf("\n");
    }
    printf("-------\n");
}

void print_sequences(char sequences[MAX_SEQS][MAX_SEQ_LEN], int num_seqs) {
    printf("Sequences loaded:\n");
    for (int i = 0; i < num_seqs; i++) {
        printf("Sequence %d: %s\n", i + 1, sequences[i]);
    }
}

// The Gibbs sampling algorithm
void gibbs_sampling(char sequences[MAX_SEQS][MAX_SEQ_LEN], int num_seqs, int k,
                    double epsilon, int starts[MAX_SEQS],
                    double pi[MAX_SEQ_LEN][4]) {
    print_sequences(sequences, num_seqs);
    double small_denom = log(num_seqs - 1 + 4 * epsilon);
    double big_denom = log(num_seqs + 4 * epsilon);

    // Initialize random starts
    for (int i = 0; i < num_seqs; i++) {
        // starts[i] = rand() % (strlen(sequences[i]) - k + 1);
        starts[i] = 0;
    }

    // Initialize probability distribution
    for (int i = 0; i < k; i++) {
        for (int j = 0; j < 4; j++) {
            pi[i][j] = epsilon;
        }
    }

    // Add motifs to the distribution
    for (int i = 0; i < num_seqs; i++) {
        add_motif(pi, sequences[i], starts[i], k, 1);
    }

    // Store the best model and likelihood
    int best_starts[MAX_SEQS];
    double best_model[MAX_SEQ_LEN][4];
    double best_likelihood = -INFINITY;

    // Perform Gibbs sampling with random sampling
    int first = 1;
    int time_since_change = 0;
    while (time_since_change < 100) {
        for (int i = 0; i < num_seqs; i++) {
            // Remove the current motif from the distribution
            add_motif(pi, sequences[i], starts[i], k, -1);

            // Calculate the probabilities for each possible start position
            double motif_distr[MAX_SEQ_LEN] = {0};
            for (int j = 0; j < (strlen(sequences[i]) - k + 1); j++) {
                motif_distr[j] =
                    motif_prob(pi, sequences[i], j, k, small_denom);
            }

            double logsum = motif_distr[0];
            for (int j = 1; j < (strlen(sequences[i]) - k + 1); j++) {
                logsum = sumLogProbs(logsum, motif_distr[j]);
            }

            // Normalize each probability distribution by subtracting the logsum
            // and exponentiating
            for (int j = 0; j < (strlen(sequences[i]) - k + 1); j++) {
                motif_distr[j] = exp(motif_distr[j] - logsum);
            }

            // Sample a new start position based on the probability distribution
            double rand_val = (double)rand() / RAND_MAX;
            double cumulative_prob = 0.0;
            int new_start = 0;
            for (int j = 0; j < (strlen(sequences[i]) - k + 1); j++) {
                cumulative_prob += motif_distr[j];
                if (rand_val <= cumulative_prob) {
                    new_start = j;
                    break;
                }
            }

            // Update the start position
            starts[i] = new_start;

            // Add the new motif back into the distribution
            add_motif(pi, sequences[i], starts[i], k, 1);
        }

        // Compute the likelihood of the new model

        double new_likelihood =
            likelihood(pi, starts, num_seqs, k, sequences, big_denom);
        if (new_likelihood > best_likelihood) {
            best_likelihood = new_likelihood;
            memcpy(best_starts, starts, num_seqs * sizeof(int));
            memcpy(best_model, pi, k * 4 * sizeof(double));
            time_since_change =
                0;  // Reset the counter since we found an improvement
        }

        time_since_change++;
    }

    // Copy the best start positions and model back to the result
    memcpy(starts, best_starts, num_seqs * sizeof(int));
    memcpy(pi, best_model, k * 4 * sizeof(double));
}

int main(int argc, char **argv) {
    srand(time(NULL));

    // Parse arguments
    char *filename = "template/motif.fasta";
    int k = 10;
    double epsilon = 0.5;

    // Read sequences
    char sequences[MAX_SEQS][MAX_SEQ_LEN];
    int num_seqs = read_fasta(filename, sequences);

    // Perform Gibbs sampling
    int starts[MAX_SEQS];
    double pi[MAX_SEQ_LEN][4];
    gibbs_sampling(sequences, num_seqs, k, epsilon, starts, pi);

    // Print results
    for (int i = 0; i < num_seqs; i++) {
        const char *sequence = sequences[i];
        int start = starts[i];
        char motif[k +
                   1];  // Buffer to hold the motif (including null terminator)

        // Extract the motif from the sequence
        strncpy(motif, &sequence[start], k);
        motif[k] = '\0';  // Null-terminate the string

        // Print the motif and its start position
        printf("Sequence %d: Start position = %d, Motif = %s\n", i, start,
               motif);
    }

    return 0;
}
